src/test/java/com/xwiki/authentication/saml/testsupport/XWikiMock.java

```diff
@@ -0,0 +1,163 @@
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.authentication.saml.testsupport;
+
+import com.xpn.xwiki.XWiki;
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.objects.StringProperty;
+import com.xpn.xwiki.objects.classes.BaseClass;
+import com.xpn.xwiki.objects.classes.StringClass;
+import com.xpn.xwiki.objects.meta.StringMetaClass;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.EntityReference;
+import org.xwiki.rendering.syntax.Syntax;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.doAnswer;
+import static org.mockito.Mockito.mock;
+
+public class XWikiMock extends XWiki {
+
+    public final Set<XWikiDocument> savedDocuments = new LinkedHashSet<>();
+    private final BaseClass groupClass;
+    private final Map<DocumentReference, XWikiDocument> docByReference = new LinkedHashMap<>();
+
+    public final BaseObject baseObjectMock = mock(BaseObject.class);
+    private final Map<String, String> savedUserAttributes = new LinkedHashMap<>();
+
+    public XWikiMock(XWikiContext context) {
+        groupClass = new BaseClass() {
+            public DocumentReference getDocumentReference() {
+                return new DocumentReference(context.getWikiId(), "XWiki:", "XWikiGroups");
+            }
+
+            @SuppressWarnings("rawtypes")
+            @Override
+            public Collection getFieldList() {
+                final Map<String, Object> fields = new LinkedHashMap<>();
+                final StringMetaClass metaClass = new StringMetaClass();
+                final StringClass prop = new StringClass("member", "XWikiGroups", metaClass);
+                fields.put("member", prop);
+
+                return fields.values();
+            }
+        };
+
+        doAnswer(invocation -> {
+            savedUserAttributes.put(invocation.getArgument(0).toString(), invocation.getArgument(1).toString());
+            return null;
+        }).when(baseObjectMock).set(any(), any(), any());
+    }
+
+    int createUserResult = 0;
+    @Override
+    public int createUser(String userName,
+                          Map<String, ?> map,
+                          EntityReference parentReference,
+                          String content,
+                          Syntax syntax,
+                          String userRights,
+                          XWikiContext context) {
+        map.forEach((key, value) -> this.savedUserAttributes.put(key + "", value + ""));
+        return createUserResult;
+    }
+
+    public void makeCreateUserReturnError(){
```

Comment:
Docs say it returns -3 on user already exists:
```suggestion
    public void makeCreateUserReturnUserAlreadyExists(){
```